File: codes/monorepo-microservice-example/user-service/src/main/java/vduczz/userservice/interfaces/controller/AuthController.java
package vduczz.userservice.interfaces.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import vduczz.userservice.application.port.in.dto.request.AuthRequestDto;
import vduczz.userservice.application.port.in.dto.response.AuthResponseDto;
import vduczz.userservice.application.service.AuthService;

@RestController
@RequestMapping("/auth")
public class AuthController {

    private final AuthService authService;
    private final AuthService authService2;

    public AuthController(
            @Qualifier("authServiceImpl")
            AuthService authService,

            @Qualifier("authServiceImpl2")
            AuthService authService2
    ) {
        this.authService = authService;
        this.authService2 = authService2;
    }

    @PostMapping("/register")
    public ResponseEntity<AuthResponseDto> register(
            @RequestBody
            @Valid
            AuthRequestDto.RegisterRequest registerRequest,

            @RequestParam(name = "mq", defaultValue = "false") boolean mq, //  use Message Queue?
            @RequestParam(name = "kafka", defaultValue = "false") boolean kafka, // Kafka event-streaming?

            @RequestParam(name = "version", defaultValue = "1") int version
    ) {

        AuthResponseDto response;
        if (version == 1)
            response = authService.register(registerRequest, mq, kafka);
        else {
            System.out.println(2);
            response = authService2.register(registerRequest, mq, kafka);
        }

        return ResponseEntity.ok(response);
    }
}

File: codes/mini-project/src/main/kotlin/com/vduczz/mini_project/controller/AuthController.kt
package com.vduczz.mini_project.controller

import com.vduczz.mini_project.dto.request.CreateUserRequest
import com.vduczz.mini_project.dto.request.LoginRequest
import com.vduczz.mini_project.dto.request.toCommand
import com.vduczz.mini_project.dto.response.AuthResponse
import com.vduczz.mini_project.service.AuthService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@Tag(name = "Auth Controller")
@RestController
@RequestMapping("/auth")
class AuthController(
    private val authService: AuthService,
) {

    @Operation(summary = "Create a new user")
    @PostMapping("/register")
    fun register(
        @Valid
        @RequestBody
        request: CreateUserRequest,
    ): ResponseEntity<AuthResponse> {
        val response = authService.register(request.toCommand())
        return ResponseEntity.ok(response)
    }

    @Operation(summary = "Log in")
    @PostMapping("/login")
    fun login(
        @Valid
        @RequestBody
        request: LoginRequest,
    ): ResponseEntity<AuthResponse> {
        val response = authService.login(request.toCommand())
        return ResponseEntity.ok(response)
    }
}